### ISSUE TITLE
test_a_star fails because a_star is nondeterministic

### ISSUE THREAD

**Issue by rcriii42** [rcriii42/Pyre#35]
Tests that pass on my Windows 10 setup with Python 3.8 fail on my Ubuntu machine with Python 3.5.

While both return the same length route, the actual route taken is different.

**Comment by rcriii42** [comment]
On Windows 10 (Note I put the route here that works on Ubuntu):

```
Failure
Traceback (most recent call last):

...

  File "C:\Users\ASUS\PycharmProjects\mPyre\src\test_a_star.py", line 46, in test_indirect_route
    self.assertListEqual(a_star((1, 1), (9, 9), test_map),

...

AssertionError: Lists differ: [(1, [20 chars](4, 4), (5, 4), (6, 4), (7, 5), (6, 6), (7, 7), (8, 8), (9, 9)] != [(1, [20 chars](4, 4), (5, 4), (6, 4), (7, 5), (7, 6), (7, 7), (8, 8), (9, 9)]

First differing element 7:
(6, 6)
(7, 6)

```

**Comment by rcriii42** [comment]
Ubuntu (note I put the route here that works on Windows):

```
Failure
Traceback (most recent call last):

  ...

  File "/home/rcriii/workspace/mPyre/src/test_a_star.py", line 47, in test_indirect_route
    [(1, 1), (2, 2), (3, 3), (4, 4), (5, 4), (6, 4), (7, 5), (6, 6), (7, 7), (8, 8), (9, 9)])

    ... 

AssertionError: Lists differ: [(1, [20 chars](4, 4), (5, 4), (6, 4), (7, 5), (7, 6), (8, 7), (9, 8), (9, 9)] != [(1, [20 chars](4, 4), (5, 4), (6, 4), (7, 5), (6, 6), (7, 7), (8, 8), (9, 9)]

First differing element 7:
(7, 6)
(6, 6)
```

**Comment by rcriii42** [comment]
Note two differences above. The first is which line the test runner (running here in pycharm) spots as the problem. In ubuntu/Py3.5 it says the problem is on line 47, in Windows/Py3.8 it says line 46. IMO line 46 is the correct answer since line 47 is a continuation of line 46.

The second difference is that in Windows/py3.8 the route goes first through square (6, 6), while in Ubuntu/Py3.5 the route passes through (7, 6).

**Comment by rcriii42** [comment]
Note the output of

```
current = min([kv for kv in f_score.items() if kv[0] in open_set], key=lambda i: i[1])[0]
print("{} fscore: {} gscore: {}".format(current,
                                                        f_score[current],
                                                       g_score[current]))

```

```
Python 3.5                          Python 3.8
-----------------------                -----------------------                 
(1, 1) fscore: 8 gscore: 0     (1, 1) fscore: 8 gscore: 0
(2, 2) fscore: 8 gscore: 1     (2, 2) fscore: 8 gscore: 1
(3, 3) fscore: 8 gscore: 2     (3, 3) fscore: 8 gscore: 2
(4, 4) fscore: 8 gscore: 3     (4, 4) fscore: 8 gscore: 3
(1, 2) fscore: 9 gscore: 1     (5, 5) fscore: 8 gscore: 4
(3, 2) fscore: 9 gscore: 2     (6, 6) fscore: 8 gscore: 5
(2, 1) fscore: 9 gscore: 1     (7, 7) fscore: 8 gscore: 6
(2, 3) fscore: 9 gscore: 2     (8, 8) fscore: 8 gscore: 7
(4, 3) fscore: 9 gscore: 3     (9, 9) fscore: 8 gscore: 8
(5, 4) fscore: 9 gscore: 4     (1, 1) fscore: 8 gscore: 0
(3, 4) fscore: 9 gscore: 3     (2, 2) fscore: 8 gscore: 1
(6, 4) fscore: 10 gscore: 5   (3, 3) fscore: 8 gscore: 2
(1, 3) fscore: 10 gscore: 2   (4, 4) fscore: 8 gscore: 3
(7, 5) fscore: 10 gscore: 6   (1, 2) fscore: 9 gscore: 1
(6, 6) fscore: 10 gscore: 7   (2, 1) fscore: 9 gscore: 1
(7, 7) fscore: 10 gscore: 8   (2, 3) fscore: 9 gscore: 2
(4, 2) fscore: 10 gscore: 3   (3, 2) fscore: 9 gscore: 2
(5, 3) fscore: 10 gscore: 4   (3, 4) fscore: 9 gscore: 3
(7, 6) fscore: 10 gscore: 7   (4, 3) fscore: 9 gscore: 3
(8, 7) fscore: 10 gscore: 8   (5, 4) fscore: 9 gscore: 4
(9, 8) fscore: 10 gscore: 9   (1, 3) fscore: 10 gscore: 2
(3, 1) fscore: 10 gscore: 2   (3, 1) fscore: 10 gscore: 2
(8, 8) fscore: 10 gscore: 9   (2, 4) fscore: 10 gscore: 3
(8, 6) fscore: 10 gscore: 7   (4, 2) fscore: 10 gscore: 3
(9, 7) fscore: 10 gscore: 8   (5, 3) fscore: 10 gscore: 4
(9, 9) fscore: 10 gscore: 10 (6, 4) fscore: 10 gscore: 5
 Here Python3.5 is done      (7, 5) fscore: 10 gscore: 6
                                             Python3.8 keeps going
```

**Comment by rcriii42** [comment]
Closed by d0b85914e525ed04499a49e1c5b127ed29f27238
